blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeImpl.java

```diff
@@ -152,16 +170,83 @@ public Link getLink() {
                                                   this);
     }
 
+    @Override
+    public boolean isRestartable()
+    {
+        RestartDeclarativePipelineAction restartDeclarativePipelineAction =
+            this.run.getAction( RestartDeclarativePipelineAction.class );
+        if ( restartDeclarativePipelineAction != null )
+        {
+            List<String> restartableStages = restartDeclarativePipelineAction.getRestartableStages();
+            if ( restartableStages != null )
+            {
+                return restartableStages.contains(this.getDisplayName())
+                    && this.getStateObj() == BlueRun.BlueRunState.FINISHED;
+            }
+        }
+        return false;
+    }
+
     @Override
     public BlueInputStep getInputStep() {
         return null;
     }
 
     @Override
     public HttpResponse submitInputStep(StaplerRequest request) {
```

Comment:
It isn't `submitInputStep`, technically it works but this doesn't look right. We should add separate method to `BluePipelineStep`?

```java
    @POST
    @WebMethod(name = "restart")
    public abstract HttpResponse restart(StaplerRequest request);
```

Reply:
LGTM good idea.
@NicuPascu please follow this coming change